:focus pseudo class fails to repaint on a shadow host when focus moves from outside the page inside a shadow tree

rniwa opened the pull request:
#### e92078a985a7af424a36ba53b36ff89412cfa4e8
<pre>
:focus pseudo class fails to repaint on a shadow host when focus moves from outside the page inside a shadow tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=205531">https://bugs.webkit.org/show_bug.cgi?id=205531</a>

Reviewed by Simon Fraser, Darin Adler and Wenson Hsieh.

The bug was caused by FrameSelection::focusedOrActiveStateChanged only invalidating the style of the focused element.
Invalidate each outer shadow host when the focus status changes to fix the bug.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::focusedOrActiveStateChanged): Fixed the bug.
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::becomeFirstResponder): Moved to UIScriptControllerIOS.
(WTR::UIScriptControllerCocoa::resignFirstResponder): Ditto.
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::becomeFirstResponder): Moved from UIScriptControllerCocoa.
(WTR::UIScriptControllerIOS::resignFirstResponder): Ditto.
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.h:
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.mm:
(WTR::UIScriptControllerMac::becomeFirstResponder): Added.
(WTR::UIScriptControllerMac::resignFirstResponder): Added.

* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-expected.html: Added.
* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.resignFirstResponder): Added the support for WebKit1.
(window.UIHelper.becomeFirstResponder): Ditto.
* LayoutTests/platform/gtk/TestExpectations: Skip the test as resignFirstResponder is not implemented on GTK+.
* LayoutTests/platform/win/TestExpectations: Ditto on Windows.

Canonical link: <a href="https://commits.webkit.org/252261@main">https://commits.webkit.org/252261@main</a>
</pre>

anttijk commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2146,11 +2146,17 @@ void FrameSelection::focusedOrActiveStateChanged()
+                break;
+            element = element->containingShadowRoot()->host();
+        } while (element);
+        if (RenderObject* renderer = element->renderer()) {
```
I don't think is the right place to do this (and we generally should not need to call invalidateStyleForSubtree at all, ever). There is already a shadow host loop in Element::setFocus. Why does that not cover this case?

rniwa replied:
The focused element isn't changing here. It's the frame / window gaining or losing the focus.

anttijk replied:
focus pseudo-class is flipping. We still need to invalidate correctly and preferably efficiently whatever the reason for flipping is.

darinadler replied:
I’m convinced!

rniwa replied:
Continuing this discussion on https://bugs.webkit.org/show_bug.cgi?id=242526.

rniwa replied:
Or this PR: https://github.com/WebKit/WebKit/pull/2259

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2146,11 +2146,17 @@ void FrameSelection::focusedOrActiveStateChanged()
+            element = element->containingShadowRoot()->host();
+        } while (element);
+        if (RenderObject* renderer = element->renderer()) {
             if (renderer && renderer->style().hasEffectiveAppearance())
```
Not new, but we’re null checking renderer twice. I suggest:

`if (auto* renderer = element->renderer(); renderer && renderer->style().hasEffectiveAppearance())`

rniwa replied:
Sure, fixed.

darinadler commented on Source/WebCore/editing/FrameSelection.cpp:
```diff
@@ -2146,11 +2146,17 @@ void FrameSelection::focusedOrActiveStateChanged()
+            element->invalidateStyleForSubtree();
+            if (!element->isInShadowTree())
+                break;
+            element = element->containingShadowRoot()->host();
```
Would it be better to call `containingShadowRoot` and null check it rather than checking `isInShadowTree` and relying on the relationship between these two checks?

darinadler replied:
Also seems like a helper function for walking up to the containing shadow root's host could make this code a lot easier to read. That idiom must have come up elsewhere.

rniwa replied:
Oh, actually we have such a function: `Node::shadowHost`. Will use that instead.

rniwa committed:
:focus pseudo class fails to repaint on a shadow host when focus moves from outside the page inside a shadow tree
https://bugs.webkit.org/show_bug.cgi?id=205531

Reviewed by Simon Fraser, Darin Adler and Wenson Hsieh.

The bug was caused by FrameSelection::focusedOrActiveStateChanged only invalidating the style of the focused element.
Invalidate each outer shadow host when the focus status changes to fix the bug.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::FrameSelection::focusedOrActiveStateChanged): Fixed the bug.
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.h:
* Tools/WebKitTestRunner/cocoa/UIScriptControllerCocoa.mm:
(WTR::UIScriptControllerCocoa::becomeFirstResponder): Moved to UIScriptControllerIOS.
(WTR::UIScriptControllerCocoa::resignFirstResponder): Ditto.
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.h:
* Tools/WebKitTestRunner/ios/UIScriptControllerIOS.mm:
(WTR::UIScriptControllerIOS::becomeFirstResponder): Moved from UIScriptControllerCocoa.
(WTR::UIScriptControllerIOS::resignFirstResponder): Ditto.
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.h:
* Tools/WebKitTestRunner/mac/UIScriptControllerMac.mm:
(WTR::UIScriptControllerMac::becomeFirstResponder): Added.
(WTR::UIScriptControllerMac::resignFirstResponder): Added.

* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host-expected.html: Added.
* LayoutTests/fast/shadow-dom/focus-ring-on-shadow-host.html: Added.
* LayoutTests/resources/ui-helper.js:
(window.UIHelper.resignFirstResponder): Added the support for WebKit1.
(window.UIHelper.becomeFirstResponder): Ditto.
* LayoutTests/platform/gtk/TestExpectations: Skip the test as resignFirstResponder is not implemented on GTK+.
* LayoutTests/platform/win/TestExpectations: Ditto on Windows.

Canonical link: https://commits.webkit.org/252261@main
